### ISSUE TITLE
Easy way to perform transfer learning

### ISSUE THREAD

**Issue by mojesty** [allenai/allennlp#2770]
**System (please complete the following information):**
 - OS: Linux
 - Python version: 3.6.1
 - AllenNLP version: 0.8
 - PyTorch version: (if you installed it yourself)

**Question**
Suppose I trained a model A (from config) that consists of several `allennlp.Module` components.
Now I want to build another model B (from similar config) where I would like to reuse weights of some components, e.g.same encoder but decoder/classification head is different, and multi loss learning is not an option.
In ideal world, I just specify path to the `best.th` of the trained model A and names of the modules I want to reuse inside B's config. 
I understand that weird things may happen due to different vocabs etc., but I want to be able to do things like that at my own risk.
How can that be done?

**Comment by joelgrus** [comment]
there is a `PretrainedModelInitializer` that is designed to initialize the weights of some of your modules with pretrained weights:

https://github.com/allenai/allennlp/blob/master/allennlp/nn/initializers.py#L193

it sounds like that's what you need

**Comment by HarshTrivedi** [comment]
There is also a transfer-modules API (#2471). Don't be misguided by the example at the top of PR, the specific API changed quite a bit later. [This](https://github.com/HarshTrivedi/allennlp/blob/2c5fbbbeda2c19d01549328c0c7edd9f6d5f4fc3/allennlp/tests/common/from_params_test.py#L275) is the test_case you should see to understand how it works. [This](https://github.com/HarshTrivedi/allennlp/blob/2c5fbbbeda2c19d01549328c0c7edd9f6d5f4fc3/allennlp/models/archival.py#L28) is where module extracting logic is implemented.

**Comment by HarshTrivedi** [comment]
TLDR, if you had original text field embedder as following 

```
"text_field_embedder": {
  "token_embedders": {
    "tokens": {
        "type": "embedding",
        "projection_dim": 2,
        "pretrained_file": "allennlp/tests/fixtures/embeddings/glove.6B.300d.sample.txt.gz",
        "embedding_dim": 300,
        "trainable": false
    }
  }
```

and you want a pretrained text_field_embedder module from somewhere, just replace that by
```
"text_field_embedder": {
    "_pretrained": {
            "archive_file": model_archive_path, 
            "module_path": "_text_field_embedder", # path to reach transfer module from pretrained model
            "freeze": True # or False
    }}
```
and you will get pretrained-module `text_field_embedder ` in your `Model` initializer.

**Comment by joelgrus** [comment]
closing, let us know if the problem is not solved

**Comment by Andy-jqa** [comment]
Hi @HarshTrivedi, @joelgrus, Thanks for you replies. I am just wondering what if I want to transfer pretrained weights of the same model to another dataset using jsonnet training config?

To be specific, I use ```training_config/esim_elmo.jsonnet``` to train on dataset 1, this will generate ```model.tar.gz``` in the output directory. Can I initialize weights in the ESIM models with the pre-trained weights in ```model.tar.gz``` and train the model on dataset 2 by just modifying the ```training_config/esim_elmo.jsonnet```?

Thanks!
Qiao

**Comment by HarshTrivedi** [comment]
If all you want is to fine-tune your trained model on a new dataset, you can use `fine-tune` command (see usage [here](https://github.com/allenai/allennlp/blob/master/allennlp/commands/fine_tune.py)).

**Comment by Andy-jqa** [comment]
> If all you want is to fine-tune your trained model on a new dataset, you can use `fine-tune` command (see usage [here](https://github.com/allenai/allennlp/blob/master/allennlp/commands/fine_tune.py)).

That is exactly what I am looking for. Thanks a lot!
